gh-102950: Implement PEP 706 – Filter for tarfile.extractall (#102953)
python/cpython: Doc/library/tarfile.rst

.. versionchanged:: 3.3
   Added support for :mod:`lzma` compression.

.. versionchanged:: 3.12
   Archives are extracted using a :ref:`filter <tarfile-extraction-filter>`,

>    which makes easy to either limit surprising/dangerous features,

Reviewer: ```suggestion
   which makes it easy to either limit surprising/dangerous features,
```

>    which makes it easy to either limit surprising/dangerous features,

Reviewer 2: I'd avoid the word "easy" and just say "possible".

   or to acknowledge that they are expected and the archive is fully trusted.

>    By default, archives are fully trusted, but this default is deprecated

Reviewer 2: Perhaps add a `:cve:` or whatever our syntax for that in ReST is to this text linking to the old tarfile cve that is in part motivating this?

Author: I don't think we have a syntax for CVEs. 
I don't want to imply that this “fixes” CVE-2007-4559, IMO it'll only be fixed in 3.14 when the default changes.

   available.

> .. method:: TarFile.extractall(path=".", members=None, *, numeric_owner=False, filter=None)

Reviewer: Any reason not to add `filter=` to `TarFile.__init__`?  We would keep the `filter=` kwarg an `.extractall` and `.extract` to override when needed.

Author: Thanks for the review! I'll get to the rest later, but here I'll reply now:

It would need to be `extraction_filter`. The `add` method has a different `filter` argument (which also isn't duplicated in `__init__`).
But I don't think putting it here would make things easier for users.

Reviewer: Ah, okay.

> .. method:: TarFile.extractall(path=".", members=None, *, numeric_owner=False, filter=None)

Reviewer 2: keyword only perhaps? (on all APIs)

Author: It is keyword-only here :)

   .. versionchanged:: 3.3
      Return an :class:`io.BufferedReader` object.

.. attribute:: TarFile.errorlevel

   If *errorlevel* is ``0``, all errors are ignored when using :meth:`TarFile.extract`
   and :meth:`TarFile.extractall`.

>    Nevertheless, they appear as error messages in the debug output, when debugging

Reviewer 2: Perhaps clarify that "when debugging is enabled" means "when debug= is greater than 0"

   is enabled.
   If ``1`` (the default), all *fatal* errors are raised as :exc:`OSError`
   exceptions. If ``2``, all *non-fatal* errors are raised as :exc:`TarError`
   exceptions as well.

   Note that when an exception is raised, the archive may be partially
   extracted. It is the user’s responsibility to clean up.

.. attribute:: TarFile.extraction_filter

   .. versionadded:: 3.12

>    The default :ref:`extraction filter <tarfile-extraction-filter>` used

Reviewer 2: "the default thing used as a default for" is redundant English-wise.  use the word default once:
"the extraction filter used as the default for the filter argument of ..." reads better

   as a default for the *filter* argument of :meth:`~TarFile.extract`
   and :meth:`~TarFile.extractall`.

>    The attribute may be ``None`` (default) or a callable.

Reviewer 2: rather than (default) here which gets a little confusing when we say it is an extraction filter above but call it None here (technically correct but confusing)... perhaps:

"The attribute may be None or a callable.  The default value is currently None but is expected to change in 3.14, see _link-to-other-place-in-docs_. Unlike extract method filter= arguments, string names are not allowed for this attribute, if not None it must be set to an implementation."

or similar?

   .. versionchanged:: 3.3
      Return an :class:`io.BufferedReader` object.

> .. attribute:: TarFile.errorlevel

Reviewer 2: add a `:type: int` ?

.. attribute:: TarInfo.name
   :type: str

   Name of the archive member.

.. attribute:: TarInfo.size
   :type: int

   Size in bytes.

.. attribute:: TarInfo.mtime
   :type: int

>    Time of last modification.

Reviewer 2: indicate what units and epoch this time integer uses (or link to an external reference if "it's complicated" rather than it represents "POSIX seconds since the 1970 UTC epoch").

Author: It actually takes float as well!
Docs for the underlying functionality, `os.stat` & `os.utime`, should mention all the quirks, so I think it's best to link. Linking to `utime` is complicated by the `ns` argument, so I went with `os.stat_result.st_mtime`.

Same for `mode` below: `os.chmod` docs have the juicy details.

   .. versionchanged:: 3.12

>       Can be set to ``None`` for :meth:`~TarFile.extract` and

Reviewer 2: for all of these "can be set to None for extract" notes, add the behavior that leads to which I _believe_ is "causing extraction to skip applying this attribute" or similarly brief wording.

      :meth:`~TarFile.extractall`.

.. attribute:: TarInfo.mode
   :type: int

>    Permission bits.

Reviewer 2: improve the wording to POSIX file permission bits perhaps?

   :type: str

   Name of the target file name, which is only present in :class:`TarInfo` objects
   of type :const:`LNKTYPE` and :const:`SYMTYPE`.

.. attribute:: TarInfo.uid
   :type: int

   User ID of the user who originally stored this member.

   .. versionchanged:: 3.12

      Can be set to ``None`` for :meth:`~TarFile.extract` and
      :meth:`~TarFile.extractall`.

.. attribute:: TarInfo.gid
   :type: int

   Group ID of the user who originally stored this member.

   .. versionchanged:: 3.12

      Can be set to ``None`` for :meth:`~TarFile.extract` and
      :meth:`~TarFile.extractall`.

.. attribute:: TarInfo.uname
   :type: str

   User name.

   .. versionchanged:: 3.12

      Can be set to ``None`` for :meth:`~TarFile.extract` and
      :meth:`~TarFile.extractall`.

.. attribute:: TarInfo.gname
   :type: str

   Group name.

   .. versionchanged:: 3.12

      Can be set to ``None`` for :meth:`~TarFile.extract` and
      :meth:`~TarFile.extractall`.

.. attribute:: TarInfo.pax_headers
   :type: dict

   A dictionary containing key-value pairs of an associated pax extended header.

.. method:: TarInfo.replace(name=..., mtime=..., mode=..., linkname=..., \

>                             uid=..., gid=..., uname=..., gname=..., \

Reviewer: Are the backslashes needed for `rst`?

Author: Turns out they aren't. Thanks!

                            deep=True)

   .. versionadded:: 3.12

   Return a *new* copy of the :class:`!TarInfo` object with the given attributes
   changed. For example, to return a ``TarInfo`` with the modification
   set to the current time, use::

>        new_tarinfo = old_tarinfo.replace(mtime=time.time())

Reviewer 2: interestingly this example might have problems...  time.time() is a float yet mtime claims to be an int.  and if someone used this example, it reveals how long processing took to extract an archive and potentially in what order based on the skew of dates on the filesystem.  people overriding mtime probably want to pick a constant up front, or if they want time.time() behavior they probably just want to set mtime=None and let the OS do that.

an example of overriding a field with None is probably simpler and more useful?

Author: Setting stuff to `None` only works for extraction. I'll use `gname='staff'`

   Return :const:`True` if it is one of character device, block device or FIFO.

.. _tarfile-extraction-filter:

Extraction filters
------------------

.. versionadded:: 3.12

> The *tar* format is designed to capture all details of a UNIX-like ecosystem,

Reviewer 3: ```suggestion
The *tar* format is designed to capture all details of a UNIX-like filesystem,
```

Reviewer 2: I also tend to say POSIX rather than UNIX at this point.

Author: POSIX is more precise, but this intentionally hand-wavy terminology: I think UNIX-*like* is more understandable to more people than POSIX-like.

* the string ``'data'``: Ignore or block most features specific to UNIX-like
  filesystems. Intended for extracting cross-platform data archives.
  See :func:`data_filter` for details.

* ``None`` (default): Use :attr:`TarFile.extraction_filter`.

  If that is also ``None`` (the default), raise a ``DeprecationWarning``,
  and fall back to the ``'fully_trusted'`` filter, whose dangerous behavior
  matches previous versions of Python.

  In Python 3.14, the ``'data'`` filter will become the default instead.
  It's possible to switch earlier; see :attr:`TarFile.extraction_filter`.

* A callable which will be called for each extracted member with a
  :ref:`TarInfo <tarinfo-objects>` describing the member and the destination
  path to where the archive is extracted (i.e. the same path is used for all
  members)::

      filter(/, member: TarInfo, path: str) -> TarInfo | None

  The callable is called just before each member is extracted, so it can
  take the current state of the disk into account.
  It can:

  - return a :class:`TarInfo` object which will be used instead of the metadata
    in the archive, or
  - return ``None``, in which case the member will be skipped, or

>   - raise an exception to abort the operation or skip the member,

Reviewer 3: I assume from the bits of the documentation above, that it should raise a TarError or OSError (or their subclasses) depending on the desired behavior with errorlevel=1. Should this be explicitly said here, if my assumption is correct?

Author: I'd rather not encourage that. `errorlevel` is rarely used and subtle; I already fixed an obvious bug in the docs (#101422). IMO the docs are technically correct and reasonably complete, and a tutorial on supporting `errorlevel` is for another time. (Though I won't write it, I'd rather deprecate `errorlevel` instead.)

    depending on :attr:`~TarFile.errorlevel`.
    Note that when extraction is aborted, :meth:`~TarFile.extractall` may leave
    the archive partially extracted. It does not attempt to clean up.

Default named filters
~~~~~~~~~~~~~~~~~~~~~

The pre-defined, named filters are available as functions, so they can be
reused in custom filters:

> .. function:: fully_trusted_filter(/, member, path)

Reviewer 3: Suggestion - add:

> Implements the ``'fully_trusted'`` filter.

>    Return *member* unchanged.

Reviewer 3: Nitpick: With "Implements" below, I would expect "Return**s**". However, disregard this comment if this is a  convention of the surrounding documentation. Same for the other filter functions and other verbs, like "strip" vs "strips".

Author: The subtle difference is intentional. The function *is* the implementation of the `'tar'` filter. When called, it will strip slashes and return a TarInfo. Two different things with slightly different grammar tenses.

    even after stripping slashes, e.g. ``C:/foo`` on Windows).
    This raises :class:`~tarfile.AbsolutePathError`.
  - :ref:`Refuse <tarfile-extraction-refuse>` to extract files whose absolute
    path (after following symlinks) would end up outside the destination.
    This raises :class:`~tarfile.OutsideDestinationError`.
  - Clear high mode bits (setuid, setgid, sticky) and group/other write bits
    (:attr:`~stat.S_IWGRP`|:attr:`~stat.S_IWOTH`).

  Return the modified ``TarInfo`` member.

.. function:: data_filter(/, member, path)

  Implements the ``'data'`` filter.
  In addition to what ``tar_filter`` does:

  - :ref:`Refuse <tarfile-extraction-refuse>` to extract links (hard or soft)
    that link to absolute paths, or ones that link outside the destination.

    This raises :class:`~tarfile.AbsoluteLinkError` or
    :class:`~tarfile.LinkOutsideDestinationError`.

    Note that such files are refused even on platforms that do not support
    symbolic links.

  - :ref:`Refuse <tarfile-extraction-refuse>` to extract device files
    (including pipes).
    This raises :class:`~tarfile.SpecialFileError`.

  - For regular files, including hard links:

>     - Set the owner read and write permissions

Reviewer 3: If the user read and write permissions are set... (continues on line 949)

      (:attr:`~stat.S_IRUSR`|:attr:`~stat.S_IWUSR`).
    - Remove the group & other executable permission
      (:attr:`~stat.S_IXGRP`|:attr:`~stat.S_IXOTH`)
      if the owner doesn’t have it (:attr:`~stat.S_IXUSR`).
    - Remove the group & other read permission
      (:attr:`~stat.S_IRGRP`|:attr:`~stat.S_IROTH`) if the owner

>       doesn’t have it (:attr:`~stat.S_IRUSR`).

Reviewer 3: (begins on line 942) ...than the owner always has it. What do I not get?

Author: Oops! Thanks for the catch!

  - For other files (directories), ignore permission bits entirely.

>   - Ignore user and group info (``uid``, ``gid``, ``uname``, ``gname``).

Reviewer 3: What does "ignore" mean in a filter? Should it be "strip"?

Reviewer 2: I'm _assuming_ it means that the code will not attempt to set the attribute on the extracted items at all. Making that more clear would be good.

  Return the modified ``TarInfo`` member.

.. _tarfile-extraction-refuse:

Filter errors
~~~~~~~~~~~~~

When a filter refuses to extract a file, it will raise an appropriate exception,
a subclass of :class:`~tarfile.FilterError`.

> This will abort the extraction if :attr:`TarFile.errorlevel` is 1 or more.

Reviewer 3: Does this mean that no builtin filter raises OSError and the behavior for errorlevel 1 and 2 is identical? Should it be explicitly mentioned?

Author: That's the goal, but I don't think it should be mentioned/guaranteed.

> With ``errorlevel=0`` the error will be logged and the member will be ignored,

Reviewer 3: ```suggestion
With ``errorlevel=0`` the error will be logged and the member will be skipped,
```

but extraction will continue.

Hints for further verification
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Even with ``filter='data'``, *tarfile* is not suited for extracting untrusted
files without prior inspection.
Among other issues, the pre-defined filters do not prevent denial-of-service
attacks. Users should do additional checks.

Here is an incomplete list of things to consider:

> * Extract to a new empty directory, to prevent e.g. exploiting pre-existing

Reviewer 2: if we're suggesting this in a security context, suggest using `tempfile.mkdtemp`.

  links, and to make it easier to clean up after a failed extraction.
* When working with untrusted data, use external (e.g. OS-level) limits on
  ·disk, memory and CPU usage.
* Check filenames against an allow-list of characters
  (to filter out control characters, confusables, foreign path separators,
  etc.).
* Check that filenames have expected extensions (discouraging files that

>   execute when you “click on them”, or extension-less files like Windows special device names),

Reviewer 3: ```suggestion
  execute when you “click on them”, or extension-less files like Windows special device names).
```

> * Limiting the number of extracted files, total size of extracted data,
>   filename length (including symlink length), size of individual files.

Reviewer: ```suggestion
* Limit the number of extracted files, total size of extracted data,
  filename length (including symlink length), and size of individual files.
```

* Check for files that would be shadowed on case-insensitive filesystems.

Also note that:

> * Tar files often contain multiple versions of the same file.

Reviewer 2: s/often/may/.

  Later ones are expected to overwrite any earlier ones.
  This feature is crucial to allow updating tape archives, but can be abused
  maliciously.
* *tarfile* does not protect against issues with “live” data,
  e.g. an attacker tinkering with the destination (or source) directory while
  extraction (or archiving) is in progress.

> Supporting older Python versions

Reviewer 2: I like that you included this section! :)

the feature.
Note that setting ``extraction_filter`` will affect any subsequent operations.

* Fully trusted archive::

    my_tarfile.extraction_filter = (lambda member, path: member)
    my_tarfile.extractall()

* Use the ``'data'`` filter if available, but revert to Python 3.11 behavior
  (``'fully_trusted'``) if this feature is not available::

    my_tarfile.extraction_filter = getattr(tarfile, 'data_filter',
                                           (lambda member, path: member))
    my_tarfile.extractall()

* Use the ``'data'`` filter; *fail* if it is not available::

    my_tarfile.extractall(filter=tarfile.data_filter)

  or::

    my_tarfile.extraction_filter = tarfile.data_filter
    my_tarfile.extractall()

* Use the ``'data'`` filter; *warn* if it is not available::

   if hasattr(tarfile, 'data_filter'):
       my_tarfile.extractall(filter='data')
   else:
       # remove this when no longer needed
       warn_the_user('Extracting may be unsafe; consider updating Python')
       my_tarfile.extractall()

Stateful extraction filter example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

While *tarfile*'s extraction methods take a simple *filter* callable,

> custom filters may be more complex objects with internal state.

Reviewer 3: ```suggestion
custom filters may be more complex objects with an internal state.
```

It may be useful to write these as context managers, to be used like this::

    with StatefulFilter() as filter_func:
        tar.extractall(path, filter=filter_func)

Such a filter can be written as, for example::

    class StatefulFilter:
        def __init__(self):
            self.file_count = 0

        def __enter__(self):
            return self

        def __call__(self, member, path):

>             self.file_count += 1
>             return member

Reviewer 3: Should the example here raise OSError/TarError when `file_count` is unreasonably large? Something like:

```suggestion
            self.file_count += 1
            if self.file_count > 1_000_000_000:
                raise OSError('This tar archive has a suspicious file count.')
            return member
```

Author: No, I don't want to imply that this is a practical filter on its own.
